### ISSUE TITLE
korean: emitting a space in an incorrect order when typing the last char with X11 applications

### ISSUE THREAD

**Issue by tsgates** [ibus/ibus#1847]
version:
  arch dist (latest, 04/03/2016), ibus:1.5.13-1, ibus-hangul: 1.5.0-2
problem: 
  "가나다 " (type: ㄱㅏㄴㅏㄷㅏ+[space]) becomes "가나 다" instead of "가나다 "

Chrome/firefox/pidgin seems working fine. But emacs (GUI or terminal), rxvt, rxvt-unicode do not work. If you have any clue (or how to fix), please let me know. It seems lots of people reported in the last (many years); this problem suddenly happened to me when I was updating ibus (say half year ago).

**Comment by fujiwarat** [comment]
We have reverted the change for ibus-x11 since it had caused Jave applications hang up for a few years [1].
I'd suggest to use GTK applications likes gnome-terminal, xfce4-terminal instead while we don't have the better fix.

Or you can restart ibus with the following steps:
% ibus exit
% env IBUS_ENABLE_SYNC_MODE=0 ibus-daemon --xim &

[1] https://github.com/ibus/ibus/issues/1713

**Comment by tsgates** [comment]
it seems fixing the problem in all applications that I am using for now. Any caveat of using " IBUS_ENABLE_SYNC_MODE=0"? A few people reported incorrectly ordered keyboard events, but at least, it seems working fine to me (rxvt, emacs in xmonad).

**Comment by fujiwarat** [comment]
IBUS_ENABLE_SYNC_MODE=0 causes freeze of the Java apps.

**Comment by Queuecumber** [comment]
Is there a way to keep `IBUS_ENABLE_SYNC_MODE=0` at startup? (I hate java and don't use it so I don't mind making this permanent)

**Comment by Queuecumber** [comment]
Actually this doesn't seem to be fixing the problem for me, at least not in chromium browser. No matter how I try to set that environment variable to 0 it seems that I am still getting these extra spaces inserted. Other applications, like gedit, do seem to function properly.

**Comment by fujiwarat** [comment]
> Is there a way to keep IBUS_ENABLE_SYNC_MODE=0 at startup? 

It depends on how you run ibus-daemon. You could create a new script which runs 'env IBUS_ENABLE_SYNC_MODE=0 ibus-daemon --xim'.

> Actually this doesn't seem to be fixing the problem for me

You have to install ibus gtk2 module besides ibus gtk3 one and have to run ibus-daemon with '--xim' option.

**Comment by Queuecumber** [comment]
Ok so to make sure I understand, chromium is using ibus for gtk2 so I need to make sure that both daemons are launched with the correct option?

**Comment by fujiwarat** [comment]
ibus gtk2 is not a daemon but a library for gtk2 applications.

**Comment by Queuecumber** [comment]
> It depends on how you run ibus-daemon. You could create a new script which runs 'env IBUS_ENABLE_SYNC_MODE=0 ibus-daemon --xim'.

I am just using Gnome 3 which seems to launch it automatically at some point, do you know how to control the launch under Gnome 3?

**Comment by Queuecumber** [comment]
So I think I already had ibus gtk2 installed but I'm not sure, on Ubuntu 16.04 I see

```
% sudo apt install ibus-gtk                                                                                                                                                                                                                                                                                                   ~ mehrlich@lilith
[sudo] password for mehrlich: 
Reading package lists... Done
Building dependency tree       
Reading state information... Done
ibus-gtk is already the newest version (1.5.11-1ubuntu2).
0 upgraded, 0 newly installed, 0 to remove and 24 not upgraded.
 % sudo apt install ibus-gtk3                                                                                                                                                                                                                                                                                                  ~ mehrlich@lilith
Reading package lists... Done
Building dependency tree       
Reading state information... Done
ibus-gtk3 is already the newest version (1.5.11-1ubuntu2).
0 upgraded, 0 newly installed, 0 to remove and 24 not upgraded.
```

So assuming the environment variable is being set properly, which I believe it is, maybe I am not launching with the `--xim` option?

**Comment by fujiwarat** [comment]
Probably it would be a dbus issue.
E.g. https://bugzilla.gnome.org/show_bug.cgi?id=738205

gnome-shell runs ibus-daemon using dbus but the latest dbus does not expose the environment variables.

The workaround might be to rename '/usr/bin/ibus-daemon' and new script '/usr/bin/ibus-daemon' would run the renamed one with IBUS_ENABLE_SYNC_MODE=0.
gnome-shell has the hard-coded name 'ibus-daemon'.

**Comment by fujiwarat** [comment]
> maybe I am not launching with the --xim option?

gnome-shell runs ibus-daemon with --xim option.

**Comment by Queuecumber** [comment]
> The workaround might be to rename '/usr/bin/ibus-daemon' and new script '/usr/bin/ibus-daemon' 

Ok i thought of this but it struck me as the "nuclear option" so I was looking for other ways, I'll give it a shot tomorrow and let you know how it goes

**Comment by fujiwarat** [comment]
You could check /proc/$PID/environ file for which environment variables are set in your application.

**Comment by fujiwarat** [comment]
DBus is used for two spots.
gnome-shell -- DBus --> ibus-daemon -- DBus --> /usr/libexec/ibus-x11
So the workaround needs to be applied to ibus-x11 too besides ibus-daemon.

**Comment by Queuecumber** [comment]
Ok more weirdness, as I sat down to apply this fix I decided to examine the environment of the `ibus-daemon` and some other processes, namely chromium which is what I was having trouble with. 

```
% pgrep ibus-daemon
21028
% cat /proc/21028/environ
```

And I see that `IBUS_ENABLE_SYNC_MODE=0` which I assume is because I am setting it in my `.profile`. This is the case for all the processes I checked including chromium which is currently the one giving me trouble, and it makes sense since after making that change, applications like gedit no longer have this issue. Based on this I think it is something going wrong specifically with ibus and chromium interacting. 

Maybe I should also be filing a bug with chromium because I am not really sure how to debug it at this point, if you have any ideas let me know.

**Comment by fujiwarat** [comment]
You can check if im-ibus.so is loaded correctly.
% strace /opt/google/chrome/chrome 2>&1 |  grep immodules
And also you can check GTK_IM_MODULE in /proc/$PID/environ

**Comment by Queuecumber** [comment]
```
% strace chromium-browser 2>&1 | grep immodules
open("/usr/lib/x86_64-linux-gnu/gtk-2.0/2.10.0/immodules.cache", O_RDONLY) = 17
stat("/usr/lib/x86_64-linux-gnu/gtk-2.0/2.10.0/immodules/im-xim.so", {st_mode=S_IFREG|0644, st_size=31560, ...}) = 0
open("/usr/lib/x86_64-linux-gnu/gtk-2.0/2.10.0/immodules/im-xim.so", O_RDONLY|O_CLOEXEC) = 17
```

I don't know if that makes sense or if I called it right for chromium instead of chrome

```
GTK_IM_MODULE=xim
CLUTTER_IM_MODULE=xim
```

**Comment by fujiwarat** [comment]
GTK_IM_MODULE should be 'ibus'.

**Comment by Queuecumber** [comment]
Ok so I should try adding that to my .profile or is chromium overriding it?

**Comment by fujiwarat** [comment]
> Ok so I should try adding that to my .profile or is chromium overriding it?

Anyway should be fine.

% env GTK_IM_MODULE=ibus  chromium-browser

**Comment by Queuecumber** [comment]
`GTK_IM_MODULE=ibus chromium-browser` seems to fix it (at least I haven't seen the problem at all while testing where it was constantly a problem before).

So the question remains why isnt this being set automatically and is chromium overriding it?

**Comment by fujiwarat** [comment]
> So the question remains why isnt this being set automatically and is chromium overriding it?

chromium should not change it but might be a failure in your setting.

**Comment by Queuecumber** [comment]
Seems weird that it wouldn't be set right by default but adding that environment variable to my .profile has made the change permanent. Thanks for the help

**Comment by crocket** [comment]
I use `st` and `urxvt` terminals that do not support GTK.
Those applications use XIM, and this issue has been affecting me.

**Comment by crocket** [comment]
@fujiwarat setting GTK_IM_MODULE and QT_IM_MODULE to xim is fine as long as `IBUS_ENABLE_SYNC_MODE` is set to `0`. I haven't tested a java GUI application.

What's wrong with XIM? XIM is necessary for X applications that don't depend on Gtk or Qt.

What's wrong with java GUI applications?

**Comment by fujiwarat** [comment]
@crocket 
issues #1713 

> I use st and urxvt terminals that do not depend on GTK.
Why don't you use GTK or QT applications?

**Comment by crocket** [comment]
@fujiwarat 

> Why don't you use GTK or QT applications?

1) There are reasons to prefer applications with minimal dependencies.
Since their dependency tree is small, they tend to be stabler and faster. I am on gentoo, and st compiles blazingly fast. st is minimal but as useful as gnome-terminal.

2) urxvt and st allow me to apply customized color themes efficiently by textual config settings.
If I wanted to apply a specific color theme that I downloaded from the internet to urxvt or st, I could just copy and paste blocks of text. Often, textual settings make it easier to apply customized settings than GUI settings do. After all, **configuration** is **a trimmed-down version of programming**. Programming in GUI can be both easy and painful. Think about LabVIEW. I used LabVIEW in the past. LabVIEW makes certain cases easier but makes advanced programming more difficult.

Also, urxvt and st provide me with a lot of customization options that gnome-terminal, lxterminal, and xfce4-terminal do not offer.

In summary, it is easier to make urxvt and st do what I want when I want to do specific things.

**Comment by fujiwarat** [comment]
We think this problem and finally accept ForwardKeyEvent is one of the best solutions.
Recently also fixed [ForwardKeyEvent](https://codereview.qt-project.org/#/c/212179/) in Qt5 applications.

ForwardKeyEvent was implemented in ibus-hangul git but the tarball is not released yet.
I asked [ibus-hangul maintainer](https://github.com/choehwanjin/ibus-hangul/issues/52) to release a new version to include the latest git.
